applications/txn_plant/targets/freertos.armv7m.ek-tm4c123gxl/HwInit.cxx

```diff
@@ -0,0 +1,425 @@
+  static const size_t Q_SIZE = 4;
+
+
+  // Pins defined for railcom
+  using RAILCOM_TRIGGER_Pin = ::RAILCOM_TRIGGER_Pin;
+  static const auto RAILCOM_TRIGGER_INVERT = true;
+  static const auto RAILCOM_TRIGGER_DELAY_USEC = 6;
+
+  static const auto RAILCOM_UART_BASE = UART1_BASE;
+  static const auto RAILCOM_UART_PERIPH = SYSCTL_PERIPH_UART1;
+  using RAILCOM_UARTPIN = ::RAILCOM_CH1_Pin;
+};
+
+extern MCP23017 the_port;
+
+static TivaDCC<DccHwDefs> tivaDCC("/dev/mainline", &railcom_driver);
+
+extern "C" {
+/** Blink LED */
+uint32_t blinker_pattern = 0;
+static volatile uint32_t rest_pattern = 0;
+
+void dcc_generator_init(void);
+
+void hw_set_to_safe(void)
+{
+    tivaDCC.disable_output();
+    GpioInit::hw_set_to_safe();
+}
+
+void resetblink(uint32_t pattern)
+{
+    blinker_pattern = pattern;
+    /* make a timer event trigger immediately */
+}
+
+void setblink(uint32_t pattern)
+{
+    resetblink(pattern);
+}
+
+static uint32_t nsec_per_clock = 0;
+/// Calculate partial timing information from the tick counter.
+long long hw_get_partial_tick_time_nsec(void)
+{
+    volatile uint32_t * tick_current_reg = (volatile uint32_t *)0xe000e018;
+    long long tick_val = *tick_current_reg;
+    tick_val *= nsec_per_clock;
+    long long elapsed = (1ULL << NSEC_TO_TICK_SHIFT) - tick_val;
+    if (elapsed < 0) elapsed = 0;
+    return elapsed;
+}
+
+void timer5a_interrupt_handler(void)
+{
+    //
+    // Clear the timer interrupt.
+    //
+    MAP_TimerIntClear(TIMER5_BASE, TIMER_TIMA_TIMEOUT);
+    // Set output LED.
+    BLINKER_RAW_Pin::set((rest_pattern & 1));
+
+    // Shift and maybe reset pattern.
+    rest_pattern >>= 1;
+    if (!rest_pattern)
+        rest_pattern = blinker_pattern;
+}
+
+void uart1_interrupt_handler(void)
+{
+  railcom_driver.os_interrupt_handler();
+}
+
+void diewith(uint32_t pattern)
+{
+    vPortClearInterruptMask(0x20);
+    asm("cpsie i\n");
+
+    resetblink(pattern);
+    while (1)
+        ;
+}
+
+/** Initialize the processor hardware.
+ */
+void hw_preinit(void)
+{
+    /* Globally disables interrupts until the FreeRTOS scheduler is up. */
+    asm("cpsid i\n");
+
+    nsec_per_clock = 1000000000 / cm3_cpu_clock_hz;
+    
+    //
+    // Unlock PF0 so we can change it to a GPIO input
+    // Once we have enabled (unlocked) the commit register then re-lock it
+    // to prevent further changes.  PF0 is muxed with NMI thus a special case.
+    //
+    MAP_SysCtlPeripheralEnable(SW2_Pin::GPIO_PERIPH);
+    HWREG(SW2_Pin::GPIO_BASE + GPIO_O_LOCK) = GPIO_LOCK_KEY;
+    HWREG(SW2_Pin::GPIO_BASE + GPIO_O_CR) |= 0x01;
+    HWREG(SW2_Pin::GPIO_BASE + GPIO_O_LOCK) = 0;
+
+    // Initializes all GPIO and hardware pins.
+    GpioInit::hw_init();
+
+    /* Setup the system clock. */
+    MAP_SysCtlClockSet(SYSCTL_SYSDIV_2_5 | SYSCTL_USE_PLL | SYSCTL_OSC_MAIN |
+                       SYSCTL_XTAL_16MHZ);
+
+    MAP_SysCtlPeripheralEnable(SYSCTL_PERIPH_I2C1);
+    MAP_GPIOPinConfigure(GPIO_PA6_I2C1SCL);
```

Comment:
This line and the next four can be replaced by an entry in the gpio pin initializer in hardware.hxx:
see example 
```
GPIO_HWPIN(BOOSTER_H, GpioHwPin, B, 6, T0CCP0, Timer);
```
so you would use something like
```
GPIO_HWPIN(I2C_SCL, GpioHwPin, A, 6, I2C1SCL, I2CSCL);
GPIO_HWPIN(I2C_SDA, GpioHwPin, A, 7, I2C1SDA, I2C);
```
then add I2C_SDA_Pin and I2C_SCL_Pin into the gpioinit array.

The reason is twofold:
- consistency with the other declarations
- it makes sure not to forget any of the pin setup calls. For example here there is a missing call to enable the clock to the gpio peripheral; you were lucky someone else did that earlier but if this was not on port A but say on port C, you would be getting a hard fault here and scratching your head for a while.